llm4pddl/main.py

```diff
@@ -1,15 +1,98 @@
 """Main script for running experiments."""
 
+import logging
+import os
+import pickle
+import sys
+import time
+from pathlib import Path
+
+from llm4pddl import utils
+from llm4pddl.approaches import create_approach
+from llm4pddl.approaches.base_approach import BaseApproach
+from llm4pddl.envs import create_env
+from llm4pddl.envs.base_env import BaseEnv
 from llm4pddl.flags import FLAGS, parse_flags
+from llm4pddl.structs import Metrics
 
 
 def _main() -> None:
     """The main entry point."""
+    script_start = time.time()
+    str_args = " ".join(sys.argv)
+    # Parse command-line flags.
     parse_flags()
-    env = FLAGS.env
-    approach = FLAGS.approach
-    print("Env:", env)
-    print("Approach:", approach)
+    # Set up logging.
+    logging.basicConfig(level=FLAGS.loglevel,
+                        format="%(message)s",
+                        handlers=[logging.StreamHandler()])
+    logging.info(f"Running command: python {str_args}")
+    logging.info("Full config:")
+    logging.info(FLAGS)
+    logging.info(f"Git commit hash: {utils.get_git_commit_hash()}")
+    # Create the approach and env.
+    approach = create_approach(FLAGS.approach)
+    env = create_env(FLAGS.env)
+    # Run the pipeline.
+    _run_pipeline(approach, env)
+    script_time = time.time() - script_start
+    logging.info(f"\n\nMain script terminated in {script_time:.5f} seconds")
+
+
+def _run_pipeline(approach: BaseApproach, env: BaseEnv) -> None:
+    """Run training and evaluation."""
+    # Run training for learning-based approaches.
+    if approach.is_learning_based:
+        train_tasks = env.get_train_tasks()
+        approach.train(train_tasks)
+    # Run evaluation for all approaches.
+    eval_tasks = env.get_eval_tasks()
+    results = _run_evaluation(approach, eval_tasks)
+    # Save the results.
+    os.makedirs(FLAGS.results_dir, exist_ok=True)
+    outdata = {
+        "config": FLAGS,
+        "results": results.copy(),
+        "git_commit_hash": utils.get_git_commit_hash()
+    }
+    outfile = Path(FLAGS.results_dir) / f"{utils.get_config_path_str()}.pkl"
```

Comment:
for the outfile definition line, what is happening with the `/` here between the Path and the string?

Reply:
good question, that's a fancy python3 feature for joining together paths: https://www.lucasmelin.com/how-does-pathlib-combine-paths-using-slashes